### DISCUSSION TITLE
Errors when try to build Nextjs app router application

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#61072] [App Router]
E:\updated-works\v2.0\seo-blog-dashboard>npm run build

> nextjs-dashboard-template@0.1.0 build
> next build

 ✓ Creating an optimized production build    
 ✓ Compiled successfully
   Skipping linting
 ✓ Checking validity of types    
 ✓ Collecting page data    
   Generating static pages (13/30) [  ==]
Error loading metadata:  TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11730:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: AggregateError
      at internalConnectMultiple (node:net:1114:18)
      at afterConnectMultiple (node:net:1667:5)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'ECONNREFUSED',
    [errors]: [ [Error], [Error] ]
  }
}
Error loading metadata:  TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11730:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: AggregateError
      at internalConnectMultiple (node:net:1114:18)
      at afterConnectMultiple (node:net:1667:5)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'ECONNREFUSED',
    [errors]: [ [Error], [Error] ]
  }
}

**My look a like Server Components Where Errors Produce**
const Page = async () => {
  const { blogRouteAllMetaData } = await getBlogMetaData() ?? {};

  return (
    <>
     
    </>
  );
};

export default Page;

const getBlogMetaData = async () => {
  try {
    const res = await fetch(process.env.NEXT_PUBLIC_API_URL + `/api/blogs`, {
      cache: "no-store",
    });

    if (!res.ok) {
      throw new Error("Failed to fetch metaData");
    }
    return res.json();
  } catch (error) {
    console.log("Error loading metadata: ", error);
  }
};

**.env**
NEXT_PUBLIC_API_URL=http://localhost:3000

**Comment by icyJoseph** [comment]
Hi,

Unfortunately, this is way too common error. I am not sure where people get this examples that use `API` routes or handlers, to fetch data as we built, but you can't do that.

The short explanation is that `API` routes require a server runtime to be executed, and that does not exist during build time. That's why you get `TypeError: fetch failed`, with `ECONNREFUSED`, meaning, there was nobody to handshake the request.

This such a common error that there's well detailed FAQ, on how to troubleshoot it:

- https://nextjs-faq.com/fetch-api-in-rsc
- https://nextjs-faq.com/fetch-api-in-getserversideprops

Basically, just fetch the things directly. Whatever is inside your API route, move directly into `getBlogMetaData`. Then, if you really want `api/blogs` to be consumable by 3rd parties, use the new `getBlogMetaData`, within it. As you have it, it'll be accessible to your client, and 3rd parties during the runtime.
